Auto-discover bacnet devices in network and add them if device-profil…

Smit-Sheth opened the pull request:
…es are present for them in database

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

tonyespy commented:
@Smit-Sheth Thanks again for your contribution.  I have a number of comments for your to address.  I also want to get a second reviewer involved, as Python isn't my strongest language.

 * First, like with your other PR, I'd suggest you break this into separate commits, and maybe extract some of the changes into a separate PR.  At minimum though, I'd break this into Python changes, Java change, and property changes.

 * It looks like neither Python files you've modified have a copyright/license headers, so we'll need to fix them before we can merge your contributions.

 * When adding significant code changes (ie. more than a few lines of code) you need to add an appropriate copyright line to the file, assigning copyright either to yourself, or your company/organization.  In case of this PR, this rule applies to pretty much all of the files you've touched.

* **watcher.properties** - missing newline @ end of file.  You've also included a Dell copyright, which since this is a new file from you, isn't really appropriate.

* **FlaskController.py**:
  * Why is there a default for scan_time?  Is it an optional parameter?  If so, a comment stating this would be a good idea.
  * I don't see scan_time actually used anywhere?  That said, Python's not my strongest language, so maybe I'm missing something?
 * I don't see REST calls being made to this endpoint in the Java code (**BacNetClientImpl**?).
* **BACNetDriver.py**:
  * **_debug** should default to 0
  * You re-factored to get rid of an internal **IOCB** class.  This probably belongs in its own commit.
* **DeviceStore.java** - why did you add this new pattern matching code?  None of the other device do this.  For newly discovered devices, address and port should already have been set by the **DeviceDiscovery** class.  If this logic is really needed, I'd suggest adding a comment explaining what the pattern matching is being used for, as it's not super obvious by reading the code.
* **schedule.properties** - why did you change the frequency of the QueryAll task?  Is this directly related to device discovery?  If not, I'd suggest moving it to it's own PR.

Smit-Sheth commented:
@tonyespy I'll make the changes and push the code for review again. Just one question though what should be the format of  the copyright line. Can you give an example.

tonyespy commented:
@Smit-Sheth  File headers need to contain two things, a copyright component, and a license component.  The copyright component is pretty simple, it consists of the word "Copyright", and optional "(c)", a year (or optionally range of years), and the name of an individual or company/organization that is declaring the copyright.

Additionally the file header should also include the license associated with the code, which in this case must be Apache 2.  The EdgeX Java code embeds a more verbose license header, whereas some of our newer Go uses an SPDX license identifier.  Both are acceptable, however in your case, it's probably best to follow the style of the existing device-bacnet code.

Here's an example from device-bacnet:

```
/*******************************************************************************
 * Copyright 2016-2017 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 * @microservice: device-mqtt
 * @author: Jim White, Dell
 * @version: 1.0.0
 *******************************************************************************/
```
...and here's another header from the newer Go export code:
```

//
// Copyright (c) 2017
// Cavium
//
// SPDX-License-Identifier: Apache-2.0
//

```

Smit-Sheth committed:
Add scan REST call

This change adds scan rest call in FlaskController.py

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth committed:
Remove IOCB class implementation

This change removes implementation of IOCB class, instead imports it
from bacpypes library

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth committed:
Add discovery mechanism implementation

This change adds implementation of discovery mechanism

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth committed:
Add discovery REST call

This change adds /discovery REST call and its associated implementation

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth committed:
Split device_name to get address and port number

This change splits the device_name to get the ip address and port number
of discovered bacnet device

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth committed:
Add watcher properties file

Signed-off-by: smit sheth <smit.sheth@mobiliya.com>

Smit-Sheth commented:
@tonyespy The pattern matching in deviceStore is basically to split the ip address and port number as we get it combined. For eg we get discovered device like 192.168.10.24:47808
I have removed the scan_time from FlaskController.py. It is not required.

tonyespy reviewed (commented):
Sorry for the delay, this looks great to me, especially the way you broke up your commits.  A couple of minor comments:

 * For future reference, I don't think changing the '**@author**' tags and adding '**@updated by**' tags is really necessary.  Don't bother changing them here though.

 * **BACNetDriver.py** - I'm not well versed in Python, however the changes all seem to make sense.  The only thing I'm not sure about is the `this_application`. The only place I see `this_application` set to anything (`DiscoveryApplication`) is in `_init_backnet`, and then it's called later on using `_ss`. and `self`. which I *think* are the references to the same server instance? If so, is the top-level variable being used at all?

Finally, as I don't have any BACnet hardware to test, I assume you've tested this, and that this should for most BACnet devices?  Are there any limitations?

Smit-Sheth commented:
@tonyespy Thanks for pointing it out this_application is not needed. I'll make the required changes. There are no limitations as such.

tonyespy commented:
@Smit-Sheth Thanks for the update! One final comment...  it looks like you added a copyright header to the new file `watcher.properties`.  I checked and while our wiki does describe copyright headers [1], it doesn't provide clear guidelines for when they should be modified.  I've made an internal query about this, but in the meanwhile, it'd probably be best for you to add a Mobiliya copyright line to each of the files you've modified.  Also when you do this, please don't include a trailing ".", as a copyright line isn't a formal sentence.

[1] https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide#Contributor'sGuide-SubmissionLicenseandCopyrights

Smit-Sheth commented:
@tonyespy Removed the trailing '.' in license header. As other files are written by Dell, I have added the 'Updated by' tag in license header as suggested by James White.

tonyespy commented:
@Smit-Sheth Thanks.  I've kicked off an internal project discussion about this, and will get back to you shortly.  We need to clarify our policy with regards to copyright notices.  I'll get back to you shortly on this...

tonyespy commented:
@Smit-Sheth We discussed this internally and the consensus is that when significant changes (i.e. not 
 minor bug fixes) are made to a file, that a new copyright line may optionally be added with the date and name of developer/entity. As such I added a new note to our contributor's page in the [License & Copyrights](https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide#Contributor'sGuide-SubmissionLicenseandCopyrights) section.

To do so, you'd just add a line below the Dell copyright for each file, so for example, BACNetDriver.py would look like this:

```
# Copyright 2016-2017 Dell Inc.
# Copyright 2017-2018 Mobiliya
```

So it's your call whether you'd like to amend the commits.  If not, I'll just go ahead and merge.

Smit-Sheth commented:
@tonyespy  I have added the copyright header in python files, as significant amount of changes has been done there. You can go ahead and merge them

tonyespy reviewed (approved):
LGTM!
